pkg/monitor/clientview/gatherer.go

```diff
@@ -0,0 +1,147 @@
+package clientview
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	prometheustypes "github.com/prometheus/common/model"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	routeclient "github.com/openshift/client-go/route/clientset/versioned"
+	"github.com/openshift/library-go/test/library/metrics"
+	"github.com/openshift/origin/pkg/monitor/monitorapi"
+)
+
+func FetchEventIntervalsForRestClientError(ctx context.Context, config *rest.Config, startTime time.Time) ([]monitorapi.EventInterval, error) {
+	kubeClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	routeClient, err := routeclient.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	_, err = kubeClient.CoreV1().Namespaces().Get(ctx, "openshift-monitoring", metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return []monitorapi.EventInterval{}, nil
+	}
+	client, err := metrics.NewPrometheusClient(ctx, kubeClient, routeClient)
+	if err != nil {
+		return nil, err
+	}
+
+	gatherer := &gatherer{start: startTime, client: client}
+	result, err := gatherer.query(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("metric query for rest client failed: %v", err)
+	}
+
+	return gatherer.parse(result)
+}
+
+type gatherer struct {
+	start  time.Time
+	client prometheusv1.API
+}
+
+func (g *gatherer) query(ctx context.Context) (prometheustypes.Value, error) {
+	// TODO: should we include 5xx? "429|5..|<error>"
+	query := `rest_client_requests_total{code =~ "429|5..|<error>"}`
```

Comment:
I wonder if we can add DNS resolve error - or it won't be recorded?

Reply:
yes, when Michal's upstream PR merges, we can incorporate it.

Reply:
to be more clear, if dns error happens code=<error>, any other network error, code=<error> also. (we can't make the distinction today with rest_client_requests_total. Michal's PR adds a new metric just to capture dns error.